input_cl2.hpp

```diff
@@ -6699,31 +6699,49 @@ class CommandQueue : public detail::Wrapper<cl_command_queue>
         }
         else {
             Device device = context.getInfo<CL_CONTEXT_DEVICES>()[0];
+            bool useWithProperties;
 
-#if CL_HPP_TARGET_OPENCL_VERSION >= 200
-            cl_queue_properties queue_properties[] = {
-                CL_QUEUE_PROPERTIES, properties, 0 };
-            if ((properties & CL_QUEUE_ON_DEVICE) == 0) {
-                object_ = ::clCreateCommandQueueWithProperties(
-                    context(), device(), queue_properties, &error);
-            }
-            else {
-                error = CL_INVALID_QUEUE_PROPERTIES;
+#if CL_HPP_TARGET_OPENCL_VERSION >= 200 && CL_HPP_MINIMUM_OPENCL_VERSION < 200
+            // Run-time decision based on the actual platform
+            {
+                cl_uint version = detail::getContextPlatformVersion(context());
+                useWithProperties = (version >= 0x20000); // OpenCL 2.0 or above
             }
+#elif CL_HPP_TARGET_OPENCL_VERSION >= 200
+            useWithProperties = true;
+#else
+            useWithProperties = false;
+#endif
 
-            detail::errHandler(error, __CREATE_COMMAND_QUEUE_WITH_PROPERTIES_ERR);
-            if (err != NULL) {
-                *err = error;
+#if CL_HPP_TARGET_OPENCL_VERSION >= 200
+            if (useWithProperties) {
+                cl_queue_properties queue_properties[] = {
+                    CL_QUEUE_PROPERTIES, properties, 0 };
+                if ((properties & CL_QUEUE_ON_DEVICE) == 0) {
+                    object_ = ::clCreateCommandQueueWithProperties(
+                        context(), device(), queue_properties, &error);
+                }
+                else {
+                    error = CL_INVALID_QUEUE_PROPERTIES;
+                }
+
+                detail::errHandler(error, __CREATE_COMMAND_QUEUE_WITH_PROPERTIES_ERR);
+                if (err != NULL) {
+                    *err = error;
+                }
             }
-#else
-            object_ = ::clCreateCommandQueue(
-                context(), device(), properties, &error);
+#endif // CL_HPP_TARGET_OPENCL_VERSION >= 200
+#if CL_HPP_MINIMUM_OPENCL_VERSION < 200
```

Comment:
Hi James,

Thanks for taking the time to review this and thank you for your comments.

I think the second ``#if`` you highlighted cannot be turned into an ``#else`` (and therefore the ``if (!useProperties)`` cannot be removed either). You are right that when pointing out that ``CL_HPP_TARGET_OPENCL_VERSION < 200`` implies ``CL_HPP_MINIMUM_OPENCL_VERSION < 200``, but it's not necessarily true the other way around (and is the situation I found myself in when I found this problem, I have minimum set to 1.1, but target is 2.0); therefore they are not mutually exclusive and should be accounted for separately. That's also why the ``useProperties`` is defined in all cases and used inside the final ``#if CL_HPP_MINIMUM_OPENCL_VERSION < 200``.

I know the code is can be a bit hard to read this way. Still, if you have a look at the Image2D class constructors you will find a similar logic applied there, which is what I based my solution on.

Please let me know what do you think, and if I should give the patch another try.